/*
 * Copyright (C) EdgeTX
 *
 * Based on code named
 *   opentx - https://github.com/opentx/opentx
 *   th9x - http://code.google.com/p/th9x
 *   er9x - http://code.google.com/p/er9x
 *   gruvin9x - http://code.google.com/p/gruvin9x
 *
 * License GPLv2: http://www.gnu.org/licenses/gpl-2.0.html
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

syntax = "proto3";

package edgetx.lua.debugger;

/*
 * data types
 */

// This enum will extend in the future with
// a release of EdgeTX 3.0
enum DebugTarget {
    SCRIPT = 0;
}

enum DebuggerState {
    RUNNING = 0;
    STOPPED_BREAKPOINT = 1;
    STOPPED_ERROR = 2;
}

enum Command {
    CONTINUE = 0;
    STEP_INTO = 1;
    STEP_OVER = 2;
    PRINT = 3;
    STOP = 4;
}

message Breakpoint {
    string file = 1;
    int32 line = 2;
}

// for more info look at https://www.lua.org/manual/5.4/manual.html#lua_Debug
// this message is basically a ripoff of lua_Debug type in Lua C API
message LuaDebugInfo {
    string sourceFile = 1;
    int32 line = 2;
    What what = 3;
    optional string name = 4;
    optional NameWhat nameWhat = 5;
    bool isTailCall = 6;
    int32 nUps = 7;
    int32 nParams = 8;
    bool isVararg = 9;
    int32 fTransfer = 10;
    int32 nTransfer = 11;

    enum What {
        LUA = 0;
        C = 1;
        MAIN = 2;
    }

    enum NameWhat {
        GLOBAL = 0;
        LOCAL = 1;
        METHOD = 2;
        FIELD = 3;
        UPVALUE = 4;
    }
}

/*
 * messages
 */

// client to radio
message ExecuteExpression {
    string expression = 1;
}

// client to radio
message ExecuteDebuggerCommand {
    Command command = 1;
}

// client to radio
message SetBreakpoint {
    enum State {
        ENABLED = 0;
        DISABLED = 1;
        NONE = 2; // basically means removed
    }
    Breakpoint breakpoint = 1;
    State state = 2;
}

// client to radio
message StartDebugger {
    DebugTarget targetType = 1;
    string targetName = 2;
    repeated Breakpoint breakpoints = 3;
}

// radio to client
message DebugStatus {
    DebuggerState state = 1;
    optional Breakpoint stoppedAtBreakpoint = 2;
}

// client to radio
message GetBreakpoints {
    // If a filename is present, returned message will only
    // contain breakpoints for that filename. If not, then 
    // all breakpoints will be returned.
    optional string filename = 1;
}

// radio to client
message Breakpoints {
    optional string filename = 1;
    repeated Breakpoint breakpoints = 2;
}
